### DISCUSSION TITLE
Función como parámetro de otra función

### DISCUSSION THREAD

**Discussion by santiemanuel** [santiemanuel/milProgramadoresQA#12] [Conceptos]
Este post lo escribo como ayuda adicional para el ejercicio 22 del TP M3.

El primer paso es crear las funciones correspondientes para resolverlo.

```py
def descontar(precio, descuento):
    return precio * (1 - descuento)

def impuesto_IVA(valor):
    return valor * 1.21
```

Esto es como ya se había aprendido a escribir funciones. Ahora, en Python, si bien infiere los tipos y no hace falta indicarlos, cuenta con la posibilidad de usar `type hinting`  que nos sirve para anotar una función e indicar el tipo de dato que debería tener cada parámetro así como el tipo de dato que devolverá esta.

Traduciendo las funciones anteriores a esta alternativa nos quedan:
```py
def descontar(precio: float, descuento: float) -> float:
    return precio * (1 - descuento)

def impuesto_IVA(valor: float) -> float:
    return valor * 1.21
```

La gramática para esto sería:
```py
def funcion(param1: tipo, param2: tipo, param3: tipo) -> tipo:
    # cuerpo de la función
```

Hasta aquí tenemos las funciones individuales, cada una realizando una tarea específica. El siguiente paso es avanzar a lo solicitado, crear una función que aplique una función recibida como parámetro.

```py
def aplicar_funcion(funcion, *args) -> float:
    return funcion(*args)
```

Aquí tenemos nuevos conceptos, `aplicar_funcion` recibe como parámetro el nombre de la función (en este caso no uso type hinting para no introducir más conceptos) y `*args` que ya lo vimos en teoría.

`*args` nos sirve para pasar una cantidad variable de parámetros a una función. Dentro de esta función solo devolvemos el resultado de llamar a la función nombrada con los parámetros recibidos. Y sabemos que el resultado será `float` para ambos casos.

Ahora, primero veamos cómo se aplica todo esto.

```py
print(aplicar_funcion(descontar, 100, 0.1))
print(aplicar_funcion(impuesto_IVA, 100))
```

Se escribe el nombre de la función sin comillas, seguido de los parámetros que esta recibe separado por comas. Sabemos que `descuento` recibe 2 parámetros, entonces mandamos `100` y `0.1` luego de esta, sabemos que `impuesto_IVA` solo recibe 1 parámetro, entonces mandmos solo `100`.

Lo que ocurre por debajo es que Python reemplaza `funcion` en la llamada de `aplicar_funcion` y el resto de los parámetros separados por coma los encapsula en el segundo parámetro `*args`.

Si la función recibida recibe la misma cantidad y tipo de parámetros que encontramos en `*args`, entonces esta devolverá su resultado y a su vez `aplicar_funcion` devolverá el valor recibido por la función que se llamó. Caso contrario, recibiremos un error de tipo indicando que la función llamada recibe una cantidad diferente de parámetros.

**Comment by cdv13** [comment]
Hola santi,
si queremos pasar los valores por medio de un diccionario, seria asi?
`carro={a:(100 , 0.1) , b:(150 , 0.1 )}` -->seria el carro con el precio y descuento
`print(aplicar_funcion(descontar,carro))`
`print(aplicar_funcion(impuesto_IVA, carro))`

**Reply by santiemanuel** [reply]
Hola, las funciones escritas reciben como parámetro un valor atómico, o sea, el precio, o el descuento.

Entonces deberías crear otra función tipo `calcular_total` que reciba el carro completo (que es el diccionario). Y aquí, utilizas un ciclo para aplicar estas dos funciones a cada elemento del carro. Terminado el ciclo obtendrías el importe total en el acumulador utilizado.

**Comment by ailednitram** [comment]
Hola! A ver si estoy entendiendo bien...

```python
def discount(precio,porcentaje):
    return precio - precio * porcentaje/100

def masIva(precio, iva=21):
    return precio + precio * iva/100

carro={1000:21, 2000:10, 3000:15, 4000:20, 5000:30}

def sumar(masIva,carro):
    total=0
    for k, v in carro.items():
        subTotal=masIva(k,v)
        total+=subTotal
        print(f'Precio: {k} | Impuesto: {v} | Subtotal: {subTotal}')
    print(f'Total: {total}\n')
def descontar(discount,carro):
    total=0
    for k, v in carro.items():
        subTotal=discount(k,v)
        total+=subTotal
        print(f'Precio: {k} | Descuento: {v} | Subtotal: {subTotal}')
    print(f'Total: {total}')

sumar(masIva,carro)

descontar(discount,carro)
```

**Reply by santiemanuel** [reply]
Te faltaría tener una función que envuelva como en el ejemplo. Como está aplicado en tu código las funciones `sumar` y `descontar` les dan significados diferentes a los valores del diccionario. Si bien el ejercicio no es del todo preciso, podemos considerar al IVA como 21%, así la función envolvente se ocupe de asignar la función y los parámetros.

**Comment by ailednitram** [comment]
Ah ok, lo entendí como que tenían que ser independientes, ahí lo modifico, gracias
 
Enviado desde mi iPhone
